### PR TITLE
feat: allow any config to be defined inline

### PR DESCRIPTION
**Pull request by DaniPopes**
Allow any configuration value to be defined inline using the same TOML syntax as foundry.toml.

This is achieved by updating `Config`, `Executor` etc. fields with the new inline `Config` value saved in a `TestRunnerConfig` field for each level of tests:
- `MultiContractRunner`, global
- `ContractRunner`, contract-level, applied before running any EVM code
- `FunctionRunner` (new), function-level, applied before running each test

`TestRunnerConfig` is inherited from the level before, and updated with inline config whenever it's available.

Not *all* configs will do something at every level; for example the fork stuff is spawned at contract-level so changing that at function level won't have any effect, but for that there is cheatcodes.

`TestOptions`'s functionality has been moved to `FunctionRunner`.

Closes https://github.com/foundry-rs/foundry/issues/8564.
Closes https://github.com/foundry-rs/foundry/issues/6908.
Closes https://github.com/foundry-rs/foundry/issues/6228.

### CODE REVIEW

File: crates/forge/tests/cli/inline_config.rs [hunk 1]
```diff
@@ -147,14 +147,14 @@ forgetest!(invalid_value, |prj, cmd| {
 Compiler run successful!
 
 Ran 1 test for test/inline.sol:Inline
-[FAIL: invalid type: found sequence, expected u32 for key "default.runs.fuzz" in inline config] test(bool) ([GAS])
+[FAIL: invalid type: found sequence, expected u32 for key "default.fuzz.runs" in inline config] setUp() ([GAS])
```
Comment (DaniPopes) [inline crates/forge/tests/cli/inline_config.rs:150]: I guess `extract_inner` flipped the keys? Lol
Anyway it's gone now

Comment (grandizzy) [reply crates/forge/tests/cli/inline_config.rs:150]: :) oops, missed in review (it affects err msg only so not a biggie)

**Review by grandizzy — COMMENTED** [review]
Awesome stuff! gave https://github.com/foundry-rs/foundry/issues/6228#issuecomment-1812843644 a try with inline `paris` and `shanghai` and works great. I think this PR should close #6228 too per https://github.com/foundry-rs/foundry/issues/6228#issuecomment-1812970972 (1st and 3rd bullets are covered in this PR, the 2nd bullet / automatic detection is tracked in https://github.com/foundry-rs/foundry/issues/6440)

should we clean up tests a little bit too (like merging testdata/paris in testdata/default and removing the `TEST_DATA_PARIS` / paris `ForgeTestProfile` profile and using inline config for specifying the evm version), or better to keep separate profiles? (I can follow up in a PR after we merge this one)

**Review by grandizzy — APPROVED** [review]
lgtm!

**Review by mattsse — APPROVED** [review]
hell yeah

**Review by zerosnacks — APPROVED** [review]

**Comment by smartcontracts** [comment]
@DaniPopes should this PR make it possible to set `optimizer_runs` for a specific contract? I tried `/// forge-config: default.optimizer_runs = 5000` without any effect

**Comment by grandizzy** [comment]
@smartcontracts you may want to check  compiler profiles and compiler restrictions added in https://github.com/foundry-rs/foundry/pull/8668

**Pull request merged**